File: 721-accounts-merge/721-accounts-merge.py
class Solution:
    def accountsMerge(self, accounts: List[List[str]]) -> List[List[str]]:
        skip = set()
        
        changed = True
        while changed:
            changed = False
            
            for i in range(len(accounts)):
                if i in skip:
                    continue
                name = accounts[i][0]
                emails = accounts[i][1:]
                
                
                for j in range(1,len(accounts[i])):
                    email = accounts[i][j]
                    
                    for k in range(i+1,len(accounts)):
                        if k in skip:
                            continue
                        if email in accounts[k]:
                            changed = True
                            skip.add(k)
                            
                            emails += accounts[k][1:]
                emails = list(set(emails))
                emails.sort()
                            
                accounts[i] = [name] + emails
                

            
        return [accounts[i] for i in range(0,len(accounts)) if i not in skip]

File: 721-accounts-merge/721-accounts-merge.js
/**
 * @param {string[][]} accounts
 * @return {string[][]}
 */
var accountsMerge = function(accounts) {
    const owners = {};
    const emails = [];
    
    for(let i in accounts){
        emails[i] = []
        for(let j=1;j<accounts[i].length;j++){
            let email = accounts[i][j];
            if(!owners[email]) owners[email] = [];
            owners[email].push(i);
            emails[i].push(email);
        }
    }
    
    let seen = new Set;
    let mergedAccounts = [];
    
    for(let i in accounts){
        if(seen.has(i)) continue;
        
        // let userEmails = new Set;
        // let stack = [i];
        // while(stack.length){
        //     let current = stack.pop();
        //     if(seen.has(current)) continue;
        //     seen.add(current);
        //     for(let email of emails[current]){
        //         userEmails.add(email);
        //         stack = stack.concat(owners[email]);
        //     }
        // }
        
        let name = accounts[i].shift();
        let userEmails = new Set(accounts[i]);
        userEmails.forEach((email)=>{
            for(let owner of owners[email]){
                seen.add(owner)
                for(let email2 of emails[owner]){
                    userEmails.add(email2);
                }
            }
        });
        
        
        let emailArray = Array.from(userEmails);
        emailArray.sort((a,b)=>{
            if(a<b) return -1;
            return 1;
        });
        
        mergedAccounts.push([name,...emailArray]);
    }
    return mergedAccounts;
};